rust-runtime/aws-smithy-http-server-python/src/middleware/handler.rs

```diff
@@ -3,21 +3,64 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-//! Execute Python middleware handlers.
-use aws_smithy_http_server::{body::Body, body::BoxBody, response::IntoResponse};
-use http::Request;
-use pyo3::prelude::*;
+//! Execute pure-Python middleware handler.
 
+use aws_smithy_http_server::body::{Body, BoxBody};
+use http::{Request, Response};
+use pyo3::{exceptions::PyRuntimeError, prelude::*, types::PyFunction};
 use pyo3_asyncio::TaskLocals;
+use tower::{util::BoxService, BoxError, Service};
 
-use crate::{PyMiddlewareException, PyRequest, PyResponse};
+use crate::util::func_metadata;
 
-use super::PyFuture;
+use super::{PyMiddlewareError, PyRequest, PyResponse};
 
-#[derive(Debug, Clone, Copy)]
-pub enum PyMiddlewareType {
-    Request,
-    Response,
+// PyNextInner represents the inner service Tower layer applied to.
+type PyNextInner = BoxService<Request<Body>, Response<BoxBody>, BoxError>;
+
+// PyNext wraps inner Tower service and makes it calleble from Python.
+#[pyo3::pyclass]
+struct PyNext(Option<PyNextInner>);
+
+impl PyNext {
+    fn new(inner: PyNextInner) -> Self {
+        Self(Some(inner))
+    }
+
+    // Consumes self by taking the inner Tower service.
+    // This method would have been `into_inner(self) -> PyNextInner`
+    // but we can't do that because we are crossing Python boundary.
+    fn take_inner(&mut self) -> Option<PyNextInner> {
+        self.0.take()
+    }
+}
+
+#[pyo3::pymethods]
+impl PyNext {
+    // Calls the inner Tower service with the `Request` that is passed from Python.
+    // It returns a coroutine to be awaited on the Python to complete the call.
+    // Note that it takes wrapped objects from both `PyRequest` and `PyNext`,
+    // so after calling `next`, consumer can't access to the `Request` or
+    // can't call the `next` again, this basically emulates consuming `self` and `Request`,
+    // but since we are crossing Python boundary we can't express it in natural Rust terms.
```

Comment:
```suggestion
    // Calls the inner Tower service with the `Request` that is passed from Python.
    // It returns a coroutine to be awaited on the Python side to complete the call.
    // Note that it takes wrapped objects from both `PyRequest` and `PyNext`,
    // so after calling `next`, consumer can't access to the `Request` or
    // can't call the `next` again, this basically emulates consuming `self` and `Request`,
    // but since we are crossing the Python boundary we can't express it in natural Rust terms.
```